pkg/machines/components/vmnetworktab.jsx

```diff
@@ -26,154 +26,199 @@ import { rephraseUI, vmId } from "../helpers.js";
 import EditNICAction from './nicEdit.jsx';
 import WarningInactive from './warningInactive.jsx';
 import './nicEdit.css';
+import { vmInterfaceAddresses } from '../libvirt-dbus.js';
 
 const _ = cockpit.gettext;
 
-const VmNetworkTab = function ({ vm, dispatch, config, hostDevices, networks, onAddErrorNotification }) {
-    const id = vmId(vm.name);
+class VmNetworkTab extends React.Component {
+    constructor(props) {
+        super(props);
 
-    const nicLookupByMAC = (interfacesList, mac) => {
-        return interfacesList.filter(iface => iface.mac == mac)[0];
-    };
+        this.state = {
+            interfaceAddress: [],
```

Comment:
It feels weird for this to be state, as it's just a particular aspect of the information handed in via properties. Shouldn't `vmInterfaceAddresses()` just be called during render, or is that too expensive and needs to be cached? But what if the address changes while the VM is running? It could at least in principle happen that the `networks` property changes, but the component would never update its IP address like that.

Reply:
You 're right. If a libvirt network changes to assign to some VM static IPs for example the VM network tab should be able to reflect the change (showing the new IPs) which is not going to happen with current implementation.
I guess it doesn't cost too much to put it in render. I will update.

Reply:
So I was changing this and realized that I can't really remove the interfaceAddresses from the state. Since it's fetched from vmInterfaceaddresses which returns a promise. If I put it in render and update the state from there I will cause an infinite re-render loop.

So, to fix this not-updating issue, I can just do polling when this tab is open. Also imagine that some VM might get IPs without that the network property changed and in this case we won't update either. For example it can dynamically assigned by some non-libvirt dnsmasq dhcpd server.

Reply:
Ah, good point, so it needs to be state indeed. But polling sounds super-ugly.. So wit the current approach I figure if you have a stopped VM and start it from this Networking page, the IP would never show up, right?

It seems to me this feature should block on libvirt-dbus sending proper events for networking changes.

Reply:
Exactly. But I wouldn't block it, for Vms who got IPs before entering the network tab is will work.

virt-manager has a refresh button next to the IP address for the user to manually try to refresh it. I could add something similar if preferred.

But I would really prefer to not block it completely, since it's quite helpful feature, even for the subset of situations where is actually works :) (And in the meantime I 'll ask if it's possible to get such signals from libvirt)

Reply:
Agreed.

Reply:
Regarding the events for the IPs changes - pasting over from IRC:
(10:23:06 AM) mprivozn: kkoukiou: regarding your event question: while it probably can be technically possible, libvirt doesn't implement that; thing is, libvirt gives dnsmasq a small binary that is executed whenever there's a DHCP traffic and the binary just updates a file that is then read for virDomainInterfaceAddresses API

Reply:
Let me chime in.
Problem is that libvirt leaves handling DHCP to an external process. While we have a small binary that is run whenever there's a DHCP traffic handled by dnsmasq, I'm not sure it's run when a particular record hits its lifetime, i.e. the records only get added and the old ones are pruned only when adding a new one.
Having said that, even if libvirt did set say inotify() over the file, it would be hit only when a new IP address is assigned which would miss the point.